core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala

```diff
@@ -0,0 +1,378 @@
+
+  /**
+   * Send a message to the corresponding [[RpcEndpoint.receiveAndReply]] asynchronously.
+   * Fire-and-forget semantics.
+   *
+   * The receiver will reply to sender's [[RpcEndpoint.receive]] or [[RpcEndpoint.receiveAndReply]]
+   * depending on which one of [[RpcResponse.reply]]s is called.
+   */
+  def sendWithReply(message: Any, sender: RpcEndpointRef): Unit
+
+  /**
+   * Send a message to the corresponding [[RpcEndpoint.receiveAndReply)]] and return a `Future` to
+   * receive the reply within a default timeout.
+   */
+  def sendWithReply[T: ClassTag](message: Any): Future[T]
+
+  /**
+   * Send a message to the corresponding [[RpcEndpoint.receiveAndReply)]] and return a `Future` to
+   * receive the reply within the specified timeout.
+   */
+  def sendWithReply[T: ClassTag](message: Any, timeout: FiniteDuration): Future[T]
+
+  def toURI: URI
+}
+
+/**
+ * Represent a host with a port
+ */
+private[spark] case class RpcAddress(host: String, port: Int) {
+  // TODO do we need to add the type of RpcEnv in the address?
+
+  val hostPort: String = host + ":" + port
+
+  override val toString: String = hostPort
+}
+
+private[spark] object RpcAddress {
+
+  /**
+   * Return the [[RpcAddress]] represented by `uri`.
+   */
+  def fromURIString(uri: String): RpcAddress = {
+    val u = new java.net.URI(uri)
+    RpcAddress(u.getHost, u.getPort)
+  }
+
+  def fromSparkURL(sparkUrl: String): RpcAddress = {
+    val (host, port) = Utils.extractHostPortFromSparkUrl(sparkUrl)
+    RpcAddress(host, port)
+  }
+}
+
+/**
+ * Indicate that a new connection is established.
+ *
+ * @param address the remote address of the connection
+ */
+private[spark] case class AssociatedEvent(address: RpcAddress)
+
+/**
+ * Indicate a disconnection from a remote address.
+ *
+ * @param address the remote address of the connection
+ */
+private[spark] case class DisassociatedEvent(address: RpcAddress)
+
+/**
+ * Indicate a network error.
+ * @param address the remote address of the connection which this error happens on.
+ * @param cause the cause of the network error.
+ */
+private[spark] case class NetworkErrorEvent(address: RpcAddress, cause: Throwable)
+
+/**
+ * A callback that [[RpcEndpoint]] can use it to send back a message or failure.
+ */
+private[spark] trait RpcResponse {
+
+  /**
+   * Reply a message to the sender. If the sender is [[RpcEndpoint]], its [[RpcEndpoint.receive]]
+   * will be called.
+   */
+  def reply(response: Any): Unit
+
+  /**
+   * Reply a message to the corresponding [[RpcEndpoint.receiveAndReply]]. If you use this one to
+   * reply, it means you expect the target [[RpcEndpoint]] should reply you something.
+   *
+   * TODO better method name?
+   *
+   * @param response the response message
+   * @param sender who replies this message. The target [[RpcEndpoint]] will use `sender` to send
+   *               back something.
+   */
+  def replyWithSender(response: Any, sender: RpcEndpointRef): Unit
```

Comment:
For example, RpcEndpoint calls `sendWithReply` to send a message to RpcEndpoint B. In B's `receiveAndReply`, it may have two requirements.

1. Reply to A but don't need a reply. It means, the replied message should be sent to `receive` method of A. (reply)
2. Reply to A but need a reply. It means, the replied message should be sent to `receiveAndReply` method of A. (replyWithSender)

That's why here we need two methods.